### DISCUSSION TITLE
アノテーションの追加について

### DISCUSSION THREAD

**Discussion by Keita0920** [lit-iphone-course/lit-iphone-course#42] [Question-original] [answered]
(1) 実現したいこと
アノテーションの画像を差し替えたいけれどもアノテーション自体が表示されないので表示をしたい。

(2) 実現するために自分が試した内容とその結果
1.takePhotoメソッド以外のメソッドにも記述してみたが変化なし。
2.breakpointを使って確認したところ呼ばれてはいるがアノテーションが表示されない。
3.以下の記事に書いてあることを試してみたが変化なし。

(3) 調べた内容や資料
https://qiita.com/koogawa/items/b5d610c23d92ca4fd84e

(4) 原因だと思われそうな箇所(ソースコードをコピペして貼る)
takePhotoメソッドでピンを置いてmapViewメソッドで画像を差し替えようとしています

@IBAction func takePhoto(){
        if UIImagePickerController.isSourceTypeAvailable(.camera){
            let picker=UIImagePickerController()
            picker.sourceType = .camera
            picker.delegate=self
            picker.allowsEditing=true
            present(picker,animated: true,completion: nil)
        }else{
            print("error")
        }
        let pa = MKPointAnnotation()
        guard let loc=self.loc else{
            return
        }
        pa.coordinate = loc.coordinate
        mapView.addAnnotation(pa)
    }

func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        if (annotation is MKUserLocation) {
            // ユーザの現在地の青丸マークは置き換えない
            return nil
        } else {
            // CustomAnnotationの場合に画像を配置
            let identifier = "Pin"
            annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: identifier)
            if annotationView == nil {
                annotationView = MKAnnotationView.init(annotation: annotation, reuseIdentifier: identifier)
            }
            let tableData = realm.objects(Replace.self)
            if tableData.count == 0{
                return nil
            }
            //URL型にキャスト
            let fileURL = URL(string: tableData[0].imageURL)
            //パス型に変換
            let filePath = fileURL?.path
            annotationView?.image = UIImage(contentsOfFile: filePath!)
            annotationView?.annotation = annotation
            return annotationView
        }
    }

(5) GitHubのリンク
https://github.com/Keita0920/LA_RePlace

**Comment by txxxxc** [comment, marked as answer]
（今外なので間違ってたらごめん！）
takePhoto関数に書いてあるpresent関数はカメラ画面を表示する関数です！そのタイミングと同時にピンを刺してるので画像が反映されてないんだと思います！
didFinishPickingMadiaWithInfoが画像が選択されたタイミングで呼ばれるのでこの関数内で一旦ピンを刺す処理を書いてみて欲しいです！

ピンが刺さらなかったらまた呼んでくださいー！

**Reply by Keita0920** [reply]
//撮った写真をリサイズ
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picture=info[.editedImage]as?UIImage
        resizedPicture = picture.resize(targetSize: CGSize(width: picture.size.width / 10, height: picture.size.height / 10))
        save()
        dismiss(animated: true,completion: nil)
        let pa = MKPointAnnotation()
        guard let loc=self.loc else{
            return
        }
        pa.coordinate = loc.coordinate
        mapView.addAnnotation(pa)
    }

このimagePickerController内にピンを立てる記述をしたのですが変化がありませんでした！

**Reply by txxxxc** [reply]
guard letで弾かれてreturnされてる説と座標（coordinate）の値がおかしい説があるのでprintとかで確認してみてほしいです！！
できればコピペで共有してもらえると🙏🙏

**Reply by Keita0920** [reply]
先ほど上記のコードを書いていたviewcontrollerクラスを消して新しいクラスにコピペして使ってみたところちゃんと動いたのですが、3回ほど試したらまた画像が表示されなくなりました。

**Comment by Keita0920** [comment]
//撮った写真をリサイズ
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picture=info[.editedImage]as?UIImage
        resizedPicture = picture.resize(targetSize: CGSize(width: picture.size.width / 10, height: picture.size.height / 10))
        save()
        dismiss(animated: true,completion: nil)
        let pa = MKPointAnnotation()
        guard let loc=self.loc else{
            return
        }
        pa.coordinate = loc
        print(pa.coordinate)
        mapView.addAnnotation(pa)
    }

breakpoint機能を使ってみたところguard let以降の文も実行されていました。また、printしてみたところCLLocationCoordinate2D(latitude: 35.386948333713214, longitude: 139.5948939979543)と出てきたので値は大丈夫でした。
先ほどはlocをCLLocationで定義していましたが、今回はlocをCLLocationCoordinate2Dに変更しています。

**Reply by txxxxc** [reply]
実機で試しました！
画像が表示されない不具合は確認できなかったけど、同じ画像が表示され続けてるっぽいのでそれについて解決策一旦共有しておきます！
原因としては以下の行で毎回保存した画像データの0番目を呼び出してるのが原因だと思います！保存した画像とピンをいい感じに紐付けることが出来たらうまく表示されそうです！https://github.com/Keita0920/LA_RePlace/blob/d21724c9674053ea6f63a1187d9231231ecbb038/rePlace/HomeViewController.swift#L95

**Reply by Keita0920** [reply]
95行目の0をiにしてRealmに入っている個数分for分で回そうと考えているのですが、そもそもの写真が表示されなくなってしまっているので修正ができない状態です…

**Reply by Keita0920** [reply]
2回目のbuildをすると写真が表示されなくなってしまいます。アプリを一回消して最初のbuildの時だけ写真が表示されるようになっている状態です。この現象を解決する方法があれば教えていただきたいです。

**Reply by txxxxc** [reply]
ピンは表示されてる？画像だけ表示されてない感じ？？

**Reply by Keita0920** [reply]
ピンも表示されてないです。

**Reply by txxxxc** [reply]
2回目以降にピンが表示されていない理由は、画面が表示されたときにRealmに保存したデータを取得してそれを元にピンを刺す処理を書いていないのが原因なのでその処理をviewDidLoadなどに書いてほしいですー！

**Reply by txxxxc** [reply]
こちら解決できそうですか...?
ヤバそうだったら全然言ってくださいー！

**Reply by Keita0920** [reply]
質問していた部分までは解決しました！ありがとうございます🙇‍♂️
